File: Forth_Interpreter/Forth_Interpreter.cpp
#include "Factory.hpp"
#include "Forth_Interpreter.hpp"
#include "FuncIsNumber.hpp"
#include "Parser.hpp"

#include <fstream>
#include <iostream>
#include <stack>
#include <vector>

bool ForthInterpreter::InterpretCommand(const std::string &str) {
    std::unique_ptr<ForthCommands> executableCommand;
    std::string commandText;
    std::unique_ptr<std::string> commandsForSecondProcessing = std::make_unique<std::string>();
    bool someCout = false;

    std::vector <std::string> commands = Parser (str);

    GeneralData gendata (Stack.get(), commandText, 
                         commandsForSecondProcessing.get(), &In, &Out);

    for (unsigned int i = 0; i < commands.size(); ++i) {
        if (IsNumber(commands[i])) {
            gendata.CommandText = commands[i];
            commands[i] = "push";
        } else if (commands[i].length() > 1 && commands[i][0] == '.' && commands[i][1] == '"') {
            gendata.CommandText = commands[i];
            commands[i] = ".\"";
        } else if (commands[i].length() > 1 && commands[i][0] == 'i' && commands[i][1] == 'f') {
            gendata.CommandText = commands[i];
            commands[i] = "if";
        } else if (commands[i].length() > 1 && commands[i][0] == 'd' && commands[i][1] == 'o') {
            gendata.CommandText = commands[i];
            commands[i] = "do";
        }

        commandsForSecondProcessing->clear();
        executableCommand.reset(Factory::get().CreateInstance (commands[i]));
        executableCommand->AddGeneralData (gendata);
        someCout += executableCommand->Run();

        if (commandsForSecondProcessing != nullptr && commandsForSecondProcessing->length() > 0) {
            someCout += InterpretCommand (*commandsForSecondProcessing);
        }
    }

    return someCout;
}

ForthInterpreter::ForthInterpreter(std::istream& in, std::ostream& out) : In (in), Out (out) {}

void ForthInterpreter::InterpretString(const std::string& str) {
    Stack = std::make_unique<std::stack <int>>();
    InterpretCommand(str);
}

void ForthInterpreter::RunInterpretation(bool CinCout) {

    bool someCout;
    std::string str;
    Stack = std::make_unique<std::stack <int>>();

    if (CinCout) {
        Out << "> ";
    }

    while (getline (In, str, '\n')) {
        Out << "< ";
        someCout = InterpretCommand(str);

        if (!someCout) {
            Out << "ok";
        }
        Out << std::endl;
        if (CinCout) {
            Out << "> ";
        }
    }
}

int main(int argc, char** argv) {    
    if (argc == 1) {
        ForthInterpreter interpreter(std::cin, std::cout);
        interpreter.RunInterpretation(true);
    } else if (argc == 2) {
        std::ifstream in(argv[1]);
        ForthInterpreter interpreter(in, std::cout);
        interpreter.RunInterpretation(false);
    } else if (argc == 3) {
        std::ifstream in(argv[1]);
        std::ofstream out(argv[2]);
        ForthInterpreter interpreter(in, out);
        interpreter.RunInterpretation(false);
    } else {
        std::cout << "Undefined work mode" << std::endl;
        return -1;
    }

    return 0;
}

File: Forth_Interpreter_On_Java/src/main/java/ru/nsu/yurakov/Forth/ForthInterpreter/ForthInterpreter.java
package ru.nsu.yurakov.Forth.ForthInterpreter;

import java.io.InputStream;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.Objects;
import java.util.Scanner;
import java.util.Stack;

import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.LogManager;

import ru.nsu.yurakov.Forth.ForthCommands.ForthCommands;
import ru.nsu.yurakov.Forth.Parser.Parser;
import ru.nsu.yurakov.Forth.GeneralData.GeneralData;
import ru.nsu.yurakov.Forth.Checks.Checks;
import ru.nsu.yurakov.Forth.Factory.Factory;

public class ForthInterpreter {
    private ForthInterpreter() {}

    private ForthInterpreter (ForthInterpreter other) {}

    /** A standard constructor that sets the input and output flow
     * @param in Input stream
     * @param out Output stream
     */
    public ForthInterpreter(InputStream in, PrintStream out) {
        LOGGER.info("Was launched ForthInterpreter");
        isConsole = System.in.equals(in);
        inputFile = in;
        outputFile = out;
    }

    /** Executes the incoming string
     * @param input The line to be executed
     */
    public void InterpretString(String input) {
        LOGGER.info("Was launched InterpretString");
        sharedStack = new Stack<>();
        outputFile.print("< ");

        boolean someCout = InterpretCommand(input);

        if (!someCout) {
            outputFile.print("ok");
        }
    }

    /**
     * Starts the interpreter to work with the specified threads
     */
    public void RunInterpretation() {
        LOGGER.info("Was launched RunInterpretation");
        boolean someCout;
        String str;
        sharedStack = new Stack<>();

        if (isConsole) {
            outputFile.print("> ");
        }

        try {
            Scanner scan = new Scanner(inputFile);
            str = scan.nextLine();
            while (!str.equals("exit")) {
                outputFile.print("< ");

                someCout = InterpretCommand(str);

                if (!someCout) {
                    outputFile.print("ok");
                }
                outputFile.println();
                if (isConsole) {
                    outputFile.print("> ");
                }
                str = scan.nextLine();
            }
        } catch (Exception ex) {
            if (Objects.equals(ex.getMessage(), "No line found")) {
                return;
            }
            outputFile.println("ERROR: " + ex.getMessage());
        }
    }

    /** Executing the passed command for RunInterpretation
     * @param str The line to be executed
     * @return true if something was output to out else false
     */
    boolean InterpretCommand(String str) {
        LOGGER.info("Was launched InterpretCommand with command " + str);
        ForthCommands executableCommand = null;
        String commandText = "";
        String commandsForSecondProcessing = "";
        boolean someCout = false;

        ArrayList<String> commands;

        try {
            commands = Parser.Parsing(str);
        } catch (RuntimeException e) {
            outputFile.print("Error in Parser: " + e.getMessage());
            return true;
        }

        GeneralData gendata = new GeneralData(sharedStack, commandText, 
                            commandsForSecondProcessing, outputFile);

        for (int i = 0; i < commands.size(); ++i) {
            gendata.CommandText = commands.get(i);
            String command = gendata.CommandText.split(" ")[0];
            try {
                if (Checks.IsNumber(command)) {
                    command = "push";
                }

                executableCommand = (ForthCommands) Factory.GetFactory().CreateInstance(command);

                executableCommand.AddGeneralData(gendata);
                someCout = executableCommand.Run() || someCout;

                if (gendata.CommandsForSecondProcessing.length() > 0) {
                    someCout = InterpretCommand(gendata.CommandsForSecondProcessing) || someCout;
                }
                gendata.CommandsForSecondProcessing = "";
            } catch (ClassNotFoundException ex) {
                outputFile.print("Error in factory: " + ex.getMessage());
                return true;
            } catch (RuntimeException ex) {
                outputFile.print("Error: " + ex.getMessage());
                return true;
            }
        }
        return someCout;
    }

    /**
     * Shared stack for all instances
     */
    private Stack<Integer> sharedStack;
    /**
     * Output to the console or not
     */
    private boolean isConsole;
    /**
     * Then where will the reading be conducted from
     */
    private InputStream inputFile;
    /**
     * Where will the recording be conducted
     */
    private PrintStream outputFile;

    private static final Logger LOGGER = LogManager.getLogger(ForthInterpreter.class);
}
